Add physics for two-phase mixture flow

GiudGiud opened the pull request:
I'm fairly confident it's correctly reproducing the kernels, using the DumpObjectProblem with an upgrade I added

The doco on two phase mixture is a bit lacking now. It would be good to add a page in doc/content/modules/navier_stokes at some point.

Thoughts on first & other phase?

Do you prefer:
- first and second?
- liquid and dispersed?
- main and dispersed?
- dispersed and other?

refs #22637

GiudGiud committed:
Register the mixture physics action

GiudGiud committed:
Adapt flow physics to be more flexible with mixtures

GiudGiud committed:
Change scalar transport task registration to facilitate inheritance

GiudGiud committed:
Add physics for two component mixtures
refs #22637

GiudGiud committed:
Add two phase mixture short hands / standard terminology
refs #22637

GiudGiud committed:
Remove default BC

GiudGiud committed:
Add more getters to synchornize physics

GiudGiud committed:
Add solidifcation example with physics
refs #22637

GiudGiud committed:
Add an option to print all parameters of every object created
to debug action (physics) vs kernels syntax

GiudGiud committed:
Fix parsed functor material definition of the other phase

GiudGiud committed:
Add option to specify the liquid density separately
refs #22637

GiudGiud committed:
Work on getting a more solid convergence

GiudGiud committed:
Add density interpolation parameter to drift flux
refs #22637
Add parameter groups to mixture

GiudGiud committed:
Add test for drift flux channel w/ physics
refs #22637

GiudGiud committed:
Add support for selecting a viscosity interpolation method in flow physics
- use for stabilizing mixture models
refs #22637

GiudGiud committed:
Add support for specifying slip velocities for advection of scalar precursors
refs #22637

GiudGiud committed:
Rename confusion shorthand parameter for 2p mixture kernels
refs #22637

GiudGiud committed:
Add support for advection slip model

GiudGiud committed:
Add test for drift flux and advection slip 2-phase mixture physics
refs #22637

GiudGiud committed:
Better log for Exodus restart from Physics

GiudGiud committed:
Add routine to rename a parameter group

GiudGiud committed:
Rename scalar transport param groups for two phase mixtures
refs #22637

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2217890/) on a73d296 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27517/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27517/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 07635f237d340691516a90f7e8105a2d08e661eb`

GiudGiud committed:
Add documentation for equations solved by flow, scalar and turbulence physics

GiudGiud committed:
Add more documentation for fluid and solid energy equations

GiudGiud committed:
Add documentation for the two phase mixture physics
refs #22637

GiudGiud committed:
Define the vector friction coefficient to adapt to changes in friction term

GiudGiud committed:
Add more explicit error messages for errors with requesting/adding functors
notably if you request with one type and add after with another type, the current
error is 'you have two functors with the same name' which is confusing as one of them is really a null
created on the request

GiudGiud committed:
Add option to limit phase fraction to sensible values to
keep them realistic during a NL solve
unfortunately, which this saves bad solves, it slows down a normal one by 1 iteration so cant make that the default

GiudGiud committed:
Make INSFVMomentumDiffusion more resilient to bad viscosities

GiudGiud committed:
Add ghost layers to avoid reallocation errors on drift flux
makes no sense why 5 layers though

GiudGiud committed:
Switch up which phase to track
Add parameter to avoid Darcy vec vs scalar

GiudGiud committed:
Disallow pressure pins with pressure boundary conditions

GiudGiud committed:
Update test inputs for new physics syntax
Fix error in regular non-physics input using mixture density instead of phase fraction
Add a few parameters to make the setup work in physics

GiudGiud committed:
Only use speed functor for forchheimer

GiudGiud committed:
Further simplify inputs by including the dispersed phase drag model in the physics

GiudGiud committed:
Re-gold the mixture tests after adding missing
friction term

GiudGiud committed:
Adapt dump object problem to new syntax for input files

GiudGiud committed:
Clean up drift flux vs advection slip parameters
refs #22637

GiudGiud committed:
More rigorous choices in advection interpolation in physics

GiudGiud committed:
Bring dump objects problem up to par with my wonderful python parameter comparison script

GiudGiud committed:
Fix up lid driven input
regold.
We are still missing the friction term. It does not matter for regression testing
refs #22637

GiudGiud committed:
Make material output work for transient with mixtures
refs #22637

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2222380/) on 9967590 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27517/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27517/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 82d4457f6b164d77d42bd5a43425397e6fbc49c3`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2223147/) on 4358706 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27517/site)

This comment will be updated on new commits.

GiudGiud committed:
Fixup doco

GiudGiud committed:
Add a boolean for the solidifcation term, especially since it's inconsistent with alpha_exc

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2227418/) on 8834db3 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>26b6db</th>
      <th colspan=4>#27517 8834db</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.09%</td>
      <td>85.08%</td>
      <td>-0.01%</td>
      <td>74.58%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>103120</td>
      <td>103128</td>
      <td>+8</td>
      <td>44</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18075</td>
      <td>18084</td>
      <td>+9</td>
      <td>15</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27517/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27517/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>26b6db</th>
      <th colspan=4>#27517 8834db</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.87%</td>
      <td>83.87%</td>
      <td>-0.00%</td>
      <td>76.97%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>15983</td>
      <td>16249</td>
      <td>+266</td>
      <td>274</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3073</td>
      <td>3125</td>
      <td>+52</td>
      <td>82</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27517/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27517/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27517/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 74.58% is less than the suggested 90.0%
- `navier_stokes` new line coverage rate 76.97% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud committed:
Add rigorous parameter checking for the porous media solid phase physics

GiudGiud committed:
Add rigorous parameter checking for the two phase mixture physics

GiudGiud commented:
going to need a volunteer
@grmnptr @lindsayad @tanoret @MengnanLi91

lindsayad commented on framework/include/problems/DumpObjectsProblem.h:
```diff
@@ -77,6 +77,9 @@ class DumpObjectsProblem : public FEProblemBase
 
   std::shared_ptr<DumpObjectsNonlinearSystem> _nl_sys;
 
+  /// Whether to include all user-specified parameters in the dump
```
what happens otherwise?

GiudGiud replied:
added

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -234,8 +234,9 @@ FEProblemBase::validParams()
-      "verbose_setup", false, "Set to True to have the problem report on any object created");
+  MooseEnum verbosity("false true extra", "false");
+  params.addParam<MooseEnum>(
+      "verbose_setup", verbosity, "Set to True to have the problem report on any object created");
```
should probably add something about 'extra' to the doc-string

GiudGiud replied:
```suggestion
      "verbose_setup", verbosity, "Set to 'true' to have the problem report on any object created. Set to 'extra' to also display all parameters.");
```

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -3789,10 +3789,13 @@ FEProblemBase::swapBackMaterialsNeighbor(const THREAD_ID tid)
+                      const InputParameters & params) const
 {
-  if (_verbose_setup)
+  if (_verbose_setup == "true" || _verbose_setup == "extra")
```
```suggestion
  if (_verbose_setup != "false")
```

lindsayad commented on modules/navier_stokes/doc/content/source/physics/PNSFVSolidHeatTransferPhysics.md:
```diff
@@ -2,6 +2,39 @@
+The enthalpy is used in lieu of $\rho c_p T$ to be able to model solids with temperature dependent
+specific heat.
+
+The kernels created for this equation are:
```
```suggestion
The kernels potentially created for this equation are:
```
or just further clarify in some way that a user will not get both an isotropic and anisotropic diffusion kernel

lindsayad commented on modules/navier_stokes/doc/content/source/physics/WCNSFVFlowPhysics.md:
```diff
@@ -2,6 +2,69 @@
+For regular flow in a non-porous medium:
+
+!equation
+\nabla \cdot \rho \vec{v} = 0
```
need to include the time derivative term

lindsayad commented on modules/navier_stokes/doc/content/source/physics/WCNSFVFlowPhysics.md:
```diff
@@ -2,6 +2,69 @@
+For porous media flow:
+
+!equation
+\nabla \cdot \rho \vec{v}_D = 0
```
same

GiudGiud replied:
done

lindsayad commented on modules/navier_stokes/doc/content/source/physics/WCNSFVFlowPhysics.md:
```diff
@@ -2,6 +2,69 @@
+!alert note
+Additional details on porous media flow equations can be found on this [page](navier_stokes/pinsfv.md).
+
+The kernels created for clean flow for the mass equation:
```
```suggestion
The kernels created for free flow for the mass equation:
```

lindsayad commented on modules/navier_stokes/doc/content/source/physics/WCNSFVFlowPhysics.md:
```diff
@@ -2,6 +2,69 @@
+- [PWCNSFVMassTimeDerivative.md] for weakly-compressible flow
+- [PINSFVMassAdvection.md] for mass advection
+
+The kernels created for the momentum equation for clean flow:
```
```suggestion
The kernels created for the momentum equation for free flow:
```

lindsayad commented on modules/navier_stokes/doc/content/source/physics/WCNSFVFlowPhysics.md:
```diff
@@ -2,6 +2,69 @@
+- [INSFVMomentumDiffusion.md] for the momentum diffusion term
+- [INSFVMomentumPressure.md] for the pressure gradient term
+- [PINSFVMomentumFriction.md] for the friction term if specified
+- [INSFVMomentumGravity.md] for the gravity term
```
```suggestion
- [INSFVMomentumGravity.md] for the gravity term if specified
```

lindsayad commented on modules/navier_stokes/doc/content/source/physics/WCNSFVFlowPhysics.md:
```diff
@@ -2,6 +2,69 @@
+- [PINSFVMomentumDiffusion.md] for the momentum diffusion term
+- [PINSFVMomentumPressure.md] for the pressure gradient term
+- [PINSFVMomentumFriction.md] for the friction term if specified
+- [PINSFVMomentumGravity.md] for the gravity term
```
```suggestion
- [PINSFVMomentumGravity.md] for the gravity term if specified
```

lindsayad commented on modules/navier_stokes/doc/content/source/physics/WCNSFVFluidHeatTransferPhysics.md:
```diff
@@ -2,6 +2,57 @@
+
+This [Physics](Physics/index.md) object creates the kernels and boundary conditions to solve the advection-diffusion
+equation for the fluid temperature.
+For clean flow in a non-porous media:
```
```suggestion
For free flow in a non-porous media:
```

lindsayad commented on modules/navier_stokes/doc/content/source/physics/WCNSFVTurbulencePhysics.md:
```diff
@@ -2,6 +2,23 @@
+- [INSFVMixingLengthScalarDiffusion.md] to the scalar transport equations if the scalar transport is being solved
+
+!alert note
+These kernels are only added if each of these equations are being defined using a `Physics`.
```
```suggestion
These kernels are only added if each of these equations are being defined using their respective `Physics`.
```
I might also link to the respective physics for full clarity

lindsayad reviewed (changes requested):
14/58 of the way through

lindsayad commented on modules/navier_stokes/doc/content/source/physics/WCNSFVTwoPhaseMixturePhysics.md:
```diff
@@ -0,0 +1,60 @@
+where:
+
+- $\phi$ is the phase fraction
+- \mathbf{v} is the advecting velocity
```
```suggestion
- $\mathbf{v}$ is the advecting velocity
```

lindsayad commented on modules/navier_stokes/doc/content/source/physics/WCNSFVTwoPhaseMixturePhysics.md:
```diff
@@ -0,0 +1,60 @@
+- [INSFVScalarFieldAdvection.md] for the scalar advection term
+- [FVDiffusion.md] for the scalar diffusion term
+
+The momentum equations, if defined using a [WCNSFVFlowPhysics.md], are modified in the presence of a two-phase
+mixture. Density and viscosity should be set to their mixture values, see [#materials] for more information.
```
Also mention and link to the fluid energy physics since you discuss its modification below as well

GiudGiud replied:
done

lindsayad commented on modules/navier_stokes/doc/content/source/physics/WCNSFVTwoPhaseMixturePhysics.md:
```diff
@@ -0,0 +1,60 @@
+
+- [FVFunctorTimeKernel.md] for the time derivative for a transient solve
+- [INSFVScalarFieldAdvection.md] for the scalar advection term
+- [FVDiffusion.md] for the scalar diffusion term
```
also `NSFVPhaseChangeInterface` for the last term in the above equation right?

lindsayad commented on modules/navier_stokes/doc/content/source/physics/WCNSFVTwoPhaseMixturePhysics.md:
```diff
@@ -0,0 +1,60 @@
+- [INSFVScalarFieldAdvection.md] for the scalar advection term
+- [FVDiffusion.md] for the scalar diffusion term
+
+The momentum equations, if defined using a [WCNSFVFlowPhysics.md], are modified in the presence of a two-phase
+mixture. Density and viscosity should be set to their mixture values, see [#materials] for more information.
+
+If specified with the [!param](/Physics/NavierStokes/TwoPhaseMixture/WCNSFVTwoPhaseMixturePhysics/add_drift_flux_momentum_terms) parameter,
+a drift flux term is added to the momentum equations with the [WCNSFV2PMomentumDriftFlux.md] kernels.
+
+If specified with the [!param](/Physics/NavierStokes/TwoPhaseMixture/WCNSFVTwoPhaseMixturePhysics/add_advection_slip_term) parameter,
+an advection slip term is added to the momentum equations with the [WCNSFV2PMomentumAdvectionSlip.md] kernels.
+
+If specified with the [!param](/Physics/NavierStokes/TwoPhaseMixture/WCNSFVTwoPhaseMixturePhysics/add_phase_change_energy_term) parameter,
+a phase change source term is added to the fluid phase energy equation with the [NSFVPhaseChangeSource.md] kernel.
```
I might make these all one paragraph

GiudGiud replied:
done

lindsayad commented on modules/navier_stokes/doc/content/source/physics/WCNSFVTwoPhaseMixturePhysics.md:
```diff
@@ -0,0 +1,60 @@
+[!param](/Physics/NavierStokes/TwoPhaseMixture/WCNSFVTwoPhaseMixturePhysics/use_external_mixture_properties)
+is set to true.
+
+The gas mixture models defined in the fluid properties module cannot currently be used with
+[WCNSFVTwoPhaseMixturePhysics.md] without additional development.
```
```suggestion
The gas mixture models defined in the fluid properties module cannot currently be used by this physics
without additional development.
```

lindsayad commented on modules/navier_stokes/examples/solidification/galium_melting-physics.i:
both this file and the file it's named after misspell gallium

GiudGiud replied:
lol

lindsayad commented on modules/navier_stokes/examples/solidification/galium_melting-physics.i:
```diff
@@ -0,0 +1,225 @@
+        energy_advection_interpolation = '${advected_interp_method}'
+      []
+    []
+    [TwoPhaseMixture]
+      [mixture]
+        add_phase_transport_equation = false
+        liquid_phase_fraction_name = 'phase_1'
+        dispersed_phase_fraction_name = 'fl'
+
+        fluid_heat_transfer_physics = energy
+        add_phase_change_energy_term = true
+        use_dispersed_phase_drag_model = false
+
+        # Base phase material properties
+        first_phase_density_name = ${rho_liquid}
+        first_phase_viscosity_name = ${mu}
+        first_phase_specific_heat_name = ${cp_liquid}
+        first_phase_thermal_conductivity_name = ${k_liquid}
+        output_all_properties = true
+
+        # Other phase material properties
+        other_phase_density_name = ${rho_solid}
+        other_phase_viscosity_name = ${mu}
+        other_phase_specific_heat_name = ${cp_solid}
+        other_phase_thermal_conductivity_name = ${k_solid}
+      []
```
So I would vote to continue forward with what we have already for consistency. I would vote for `phase_1` and `phase_2`. I got really confused because `dispersed_phase_fraction_name` is pointing to the fluid fraction for which you have the dedicated parameter `liquid_phase_fraction_name` (which doesn't point to the liquid fraction??). And then later within the same parameter block we switch from a liquid-dispersed naming paradigm to first-other

GiudGiud replied:
renamed. i did this to get around the "phase 1 is an auxvariable now" and yes it was confusing. 

Works now imo

lindsayad commented on modules/navier_stokes/examples/solidification/tests:
```diff
@@ -8,4 +8,11 @@
     requirement = 'The system shall be able to simulate the melting of galium.'
     heavy = true
   []
+  [galium_melting_physics]
+    type = RunApp
+    input = 'galium_melting-physics.i'
+    cli_args = "Nx=20 Ny=10"
+    requirement = 'The system shall be able to simulate the melting of galium using the Navier Stokes physics syntax.'
+    heavy = true
+  []
```
you're just carrying-on the tradition, but it sure would be nice to test *something* other than just making sure the app runs

GiudGiud replied:
try running the original input on your machine you'll see

lindsayad commented on modules/navier_stokes/include/base/NS.h:
```diff
@@ -134,6 +134,12 @@ static const std::string c = "c";
+static const std::string latent_heat = "latent_heat";
+static const std::string T_liquidus = "T_liquidus";
+static const std::string T_solidus = "T_solidus";
+static const std::string alpha_exc = "alpha_exc";
```
I think I would like "exchange" better. "exc" makes me think of execution

GiudGiud replied:
changed

lindsayad commented on modules/navier_stokes/include/physics/WCNSFVCoupledAdvectionPhysicsHelper.h:
```diff
@@ -31,6 +31,9 @@ class WCNSFVCoupledAdvectionPhysicsHelper
   /// It is important to forward to the Physics so we do not get the smoothing status wrong
   MooseFunctorName getPorosityFunctorName(bool smoothed) const;
 
+  MooseFunctorName densityName() const { return _density_name; }
+  MooseFunctorName dynamicViscosityName() const { return _dynamic_viscosity_name; }
```
these could return const reference to avoid string copies

GiudGiud replied:
changed

lindsayad commented on modules/navier_stokes/include/physics/WCNSFVFlowPhysics.h:
```diff
@@ -50,6 +57,8 @@ class WCNSFVFlowPhysics final : public NavierStokesPhysicsBase
   const std::vector<Point> & getFluxInletDirections() const { return _flux_inlet_directions; }
   /// Get the inlet flux postprocessor if using a flux inlet
   const std::vector<PostprocessorName> & getFluxInletPPs() const { return _flux_inlet_pps; }
+  /// Get the name of the linear friction coefficient
```
Add in the doxygen that this will return something "invalid" if there's no friction

GiudGiud replied:
done

lindsayad commented on modules/navier_stokes/include/physics/WCNSFVFluidHeatTransferPhysics.h:
```diff
@@ -24,9 +24,22 @@ class WCNSFVFluidHeatTransferPhysics final : public NavierStokesPhysicsBase,
   WCNSFVFluidHeatTransferPhysics(const InputParameters & parameters);
 
+  /// Get the name of the fluid temperature variable
+  const MooseFunctorName getFluidTemperatureName() const { return _fluid_temperature_name; }
+
   /// Get the name of the specific heat material property
   const MooseFunctorName getSpecificHeatName() const { return _specific_heat_name; }
   const MooseFunctorName getSpecificEnthalpyName() const { return NS::specific_enthalpy; }
```
can return const reference

GiudGiud replied:
changed

lindsayad commented on modules/navier_stokes/include/physics/WCNSFVFluidHeatTransferPhysics.h:
```diff
@@ -24,9 +24,22 @@ class WCNSFVFluidHeatTransferPhysics final : public NavierStokesPhysicsBase,
   const MooseFunctorName getSpecificHeatName() const { return _specific_heat_name; }
   const MooseFunctorName getSpecificEnthalpyName() const { return NS::specific_enthalpy; }
+  /// Get the ambient convection parameters for parameter checking
+  std::vector<std::vector<SubdomainName>> getAmbientConvectionBlocks() const
+  {
+    return _ambient_convection_blocks;
+  }
+  /// Name of the ambient convection heat transfer coefficients for each block-group
+  std::vector<MooseFunctorName> getAmbientConvectionHTCs() const
```
const reference

GiudGiud replied:
done

lindsayad commented on modules/navier_stokes/include/physics/WCNSFVScalarTransportPhysics.h:
```diff
@@ -33,11 +39,22 @@ class WCNSFVScalarTransportPhysics final : public NavierStokesPhysicsBase,
   bool hasScalarEquations() const { return _has_scalar_equation; }
 
 protected:
+  void addFVKernels() override;
+  void addFVBCs() override;
```
```suggestion
  virtual void addFVKernels() override;
  virtual void addFVBCs() override;
```

lindsayad commented on modules/navier_stokes/include/physics/WCNSFVScalarTransportPhysics.h:
```diff
@@ -33,11 +39,22 @@ class WCNSFVScalarTransportPhysics final : public NavierStokesPhysicsBase,
+
+  /// Names of the passive scalar variables
+  std::vector<NonlinearVariableName> _passive_scalar_names;
+  /// A boolean to help compatibility with the old Modules/NavierStokesFV syntax
```
Not just that but it's necessary for your two phase mixture model physics when we're not adding the phase transport equation right?

GiudGiud replied:
added

lindsayad commented on modules/navier_stokes/include/physics/WCNSFVTwoPhaseMixturePhysics.h:
```diff
@@ -0,0 +1,82 @@
+
+  WCNSFVTwoPhaseMixturePhysics(const InputParameters & parameters);
+
+protected:
+private:
+  void addFVKernels() override;
+  void addMaterials() override;
```
```suggestion
private:
  virtual void addFVKernels() override;
  virtual void addMaterials() override;
```

lindsayad commented on modules/navier_stokes/src/fvkernels/WCNSFV2PMomentumAdvectionSlip.C:
```diff
@@ -27,6 +27,9 @@ WCNSFV2PMomentumAdvectionSlip::validParams()
   params.addParam<MooseFunctorName>("w_slip", "The slip velocity in the z direction.");
   params.addRequiredParam<MooseFunctorName>("rho_d", "Dispersed phase density.");
   params.addParam<MooseFunctorName>("fd", 0.0, "Fraction dispersed phase.");
+  params.renameParam("fd", "fraction_dispersed", "");
+  params.setDocString(NS::density, "Main phase (not dispersed) density functor");
```
Might want to rethink this doc string based off what we decide for the phase names

GiudGiud replied:
kernel syntax is in a tough spot for naming for sure with all the use cases possible. I might leave this
main & dispersed are not shocking

lindsayad commented on modules/navier_stokes/src/fvkernels/WCNSFV2PMomentumDriftFlux.C:
```diff
@@ -26,13 +26,15 @@ WCNSFV2PMomentumDriftFlux::validParams()
                              "Switch that can select face interpolation method for the density.");
 
-  params.set<unsigned short>("ghost_layers") = 2;
+  // when using the fraction of the disperse phase, somehow
+  params.set<unsigned short>("ghost_layers") = 5;
```
we're going to want to look into this

GiudGiud replied:
this is not coming from my PR either. The initial input file checked in was using density where it should have used phase fraction (see the diff on that file, iirc). once it changes it has this issue

GiudGiud replied:
I can take a look when I come back though. I was out of ideas so I would learn some new technique

GiudGiud replied:
this is fixed now

lindsayad commented on modules/navier_stokes/src/fvkernels/WCNSFV2PMomentumDriftFlux.C:
```diff
@@ -53,6 +55,11 @@ WCNSFV2PMomentumDriftFlux::WCNSFV2PMomentumDriftFlux(const InputParameters & par
         "In three dimensions, the w_slip velocity must be supplied using the 'w_slip' parameter");
+
+  // Phase fraction could be a nonlinear variable
+  const auto fraction_name = getParam<MooseFunctorName>("fraction_dispersed");
```
```suggestion
  const auto & fraction_name = getParam<MooseFunctorName>("fraction_dispersed");
```

lindsayad commented on modules/navier_stokes/src/physics/WCNSFVFlowPhysics.C:
```diff
@@ -43,6 +43,11 @@ WCNSFVFlowPhysics::validParams()
 
+  // Used for flow mixtures, where one phase is solid / not moving under the action of gravity
+  params.addParam<MooseFunctorName>(
+      "density_gravity",
```
```suggestion
      "density_for_gravity_terms",
```

lindsayad commented on modules/navier_stokes/src/physics/WCNSFVFlowPhysics.C:
```diff
@@ -1133,10 +1154,20 @@ WCNSFVFlowPhysics::addUserObjects()
+WCNSFVFlowPhysics::hasForchheimerFriction() const
+{
+  for (const auto block_i : index_range(_friction_types))
+    for (const auto & type_i : index_range(_friction_types[0]))
```
```suggestion
    for (const auto type_i : index_range(_friction_types[block_i]))
```

lindsayad commented on modules/navier_stokes/src/physics/WCNSFVFlowPhysics.C:
```diff
@@ -1349,6 +1384,51 @@ WCNSFVFlowPhysics::getPorosityFunctorName(bool smoothed) const
+  // Check all blocks. If more than one block, they would need to be consolidated #include in
+  // a single functor material. We won't implement this for now
+  if (_friction_types.empty())
+    return "0";
```
I would prefer returning an empty string unless that's going to cause issues for some reason

lindsayad replied:
Checking for an empty string is a common operation; checking for a one-character string that is "0" is less so

GiudGiud replied:
done

lindsayad commented on modules/navier_stokes/src/physics/WCNSFVFlowPhysics.C:
```diff
@@ -1349,6 +1384,51 @@ WCNSFVFlowPhysics::getPorosityFunctorName(bool smoothed) const
+    bool linear_friction_factor_found = false;
+    MooseFunctorName linear_friction_factor;
+    for (const auto block_i : index_range(_friction_types))
+      for (const auto & type_i : index_range(_friction_types[0]))
```
Is there a reason why you keep indexing at 0 instead of with `block_i`?
```suggestion
      for (const auto type_i : index_range(_friction_types[block_i]))
```

GiudGiud replied:
no it s a mistake. [0] would be fine if all blocks had the same types active

lindsayad commented on modules/navier_stokes/src/physics/WCNSFVTwoPhaseMixturePhysics.C:
```diff
@@ -0,0 +1,474 @@
+                     "Functor names for the diffusivities used for the main phase fraction.");
+
+  // Phase change parameters
+  params.addParam<MaterialPropertyName>(
```
should update this and the associated kernel to `MooseFunctorName`

GiudGiud replied:
done

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_model/lid-driven-two-phase.i:
```diff
@@ -93,6 +93,7 @@ g = -9.81
     momentum_component = 'x'
     gravity = '0 ${g} 0'
   []
+  # NOTE: the friction terms for u and v are missings
```
```suggestion
  # NOTE: the friction terms for u and v are missing
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_model/tests:
```diff
@@ -11,6 +11,16 @@
+    type = 'CSVDiff'
+    input = 'lid-driven-two-phase-physics.i'
+    csvdiff = 'lid-driven-two-phase-physics_CSV.csv'
+    requirement = 'The system shall give the correct solution for two phase flow in a lid-driven cavity with the mixture model using the Navier Stokes Physics syntax.'
```
```suggestion
    requirement = 'The system shall give the correct solution for two phase flow in a lid-driven cavity with a mixture model using Navier Stokes physics syntax.'
```
Same suggestions for other newly added requirements

lindsayad reviewed (changes requested):
finished

GiudGiud commented on framework/include/problems/DumpObjectsProblem.h:
```diff
@@ -77,6 +77,9 @@ class DumpObjectsProblem : public FEProblemBase
 
   std::shared_ptr<DumpObjectsNonlinearSystem> _nl_sys;
 
+  /// Whether to include all user-specified parameters in the dump
```
```suggestion
  /// Whether to include all user-specified parameters in the dump or only parameters that differ from the default value
```

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_model/tests:
```diff
@@ -29,6 +39,17 @@
+    type = 'Exodiff'
+    input = 'channel-drift-flux-physics.i'
+    exodiff = 'channel-drift-flux-physics_out.e'
+    requirement = 'The system shall be able to solve two-phase flow in a channel using the mixture drift flux model using the Navier Stokes Physics syntax.'
```
```suggestion
    requirement = 'The system shall be able to solve two-phase flow in a channel using a mixture drift flux model using Navier Stokes physics syntax.'
```

GiudGiud committed:
Address Alex's review
- add details on extra verbose setup
- fixup equations
- say not all terms are going to be added by default
- add virtuals to overrides
- rename alpha exchange coefficient
- use functor name for alpha coefficient
- galium -> gallium
Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2245186/) on c481a4c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27517/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27517/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ef92d30475a0cdfae7cd039c98567b33e4f620c0`

GiudGiud committed:
Rename phase fraction variables for 1 and 2

GiudGiud committed:
Rename phases in inputs

GiudGiud committed:
Add automated functor creation for cases where:
 phase 1 is set by an auxvariable
 phase 2 is not advected (for example solidification)

GiudGiud commented:
Rebased on next for the conflict. Review should be addressed

GiudGiud commented:
That distributed failure is concerning. It's on the same model as affected by the 5 ghost layer requirements
valgrind isnt saying anything
https://civet.inl.gov/job/2246126/

lindsayad commented:
a lot of sanitizer failures
[tsan.log](https://github.com/idaholab/moose/files/15418314/tsan.log)

lindsayad committed:
Use the proper thread id for moose var dependency

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_model/tests:
```diff
@@ -29,6 +39,17 @@
+    # Negative viscosity during solve
+    allow_warnings = true
+    custom_cmp = 'mixture_physics_diff.cmp'
+    min_ad_size = 80
```
we need to figure out why your physics test needs this restriction while the non-physics test doesn't

GiudGiud replied:
with the fixes, both tests now require the min ad size. it was really random that only the physics test did

GiudGiud committed:
Fix indexing in drift flux model for RC coeff computation

GiudGiud committed:
Fix an coefficient in drift flux model
Fix some typos and spelling

Update modules/navier_stokes/src/fvkernels/WCNSFV2PMomentumDriftFlux.C

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud commented:
should be ready

lindsayad commented on modules/navier_stokes/src/fvkernels/WCNSFV2PMomentumDriftFlux.C:
```diff
@@ -26,13 +26,15 @@ WCNSFV2PMomentumDriftFlux::validParams()
                              "Switch that can select face interpolation method for the density.");
 
-  params.set<unsigned short>("ghost_layers") = 2;
+  // when using the fraction of the disperse phase, somehow
+  // params.set<unsigned short>("ghost_layers") = 20;
```
```suggestion
```

GiudGiud commented:
Thanks for the review!
